### DISCUSSION TITLE
All connections in pool gradually become busy and never free up

### DISCUSSION THREAD

**Discussion by 7opf** [oracle/python-oracledb#220] [Q&A]
I run into this issue when using thin mode and not in thick mode (currently working on thick -> thin migration).

The network in our dev environment is generally unstable which causes frequent connection drops. Sometimes when this happens, a connection in the pool gets stuck in "busy" even though it is doing nothing. I have checked with conntrack and tcpdump on the network to confirm that no packets are transferred once it gets into this state.

Note that we call `connection.ping()` every 10 seconds to check whether our application is ready to serve client requests (assuming a connection can be acquired from the pool). 

We run multiple instances of our application at the same time each with their own oracle connection pool. I noticed that even though they use the same underlying network, the connections don't break at the same time. My current suspicion is that if the network interruption occurs during the ping round-trip, the connection is not cleaned up correctly (the connection is not removed from the busy list).

During the [ping interval](https://github.com/oracle/python-oracledb/blob/e861b2bbdca5ddbde5ae293e973da552116d7037/src/oracledb/impl/thin/pool.pyx#L439-L446) this cleanup happens, but there is no such cleanup in the [ping() implementation](https://github.com/oracle/python-oracledb/blob/e861b2bbdca5ddbde5ae293e973da552116d7037/src/oracledb/impl/thin/connection.pyx#L382-L385)

Is this a bug? Should I even need to ping if there's a ping_interval for every connection?

**Comment by cjbj** [comment]
What version python-oracledb?  Whether you need/want to ping would depend on how long ago you called `acquire()` and last used the connection.

**Reply by 7opf** [reply]
1.4.0.

`acquire()` is called every 10s, because every time I call `ping()` I need to acquire a connection

**Reply by 7opf** [reply]
I posted the bug report in https://github.com/oracle/python-oracledb/issues/221 following the template/questions

**Reply by cjbj** [reply]
We'll follow up in #221.
